File: hw1/src/main/scala/task2.scala
import org.apache.spark.sql.SparkSession
import org.apache.spark.SparkContext
import org.apache.spark.rdd.RDD
import org.json4s._
import org.json4s.jackson.JsonMethods._

import java.io.{File, PrintWriter}
import scala.io.Source
import scala.collection.mutable.{ArrayBuffer, ListBuffer}
import scala.math.BigDecimal


object task2 {
  implicit val formats = DefaultFormats

  def format_rdata(row: String): Tuple2[String, Float] = {
    val data = parse(row)
    return ((data \ "business_id").extract[String], (data \ "stars").extract[Float])
  }

  def format_bdata(row: String): Tuple2[String, String] = {
    val data = parse(row)
    return ((data \ "business_id").extract[String], (data \ "categories").extract[String])
  }

  def split_categories(row: Tuple2[String, String]): List[Tuple2[String, String]] = {
    var out = new ListBuffer[Tuple2[String, String]]()
    for (x <- row._2.trim().split(",")) {
      out += ((row._1, x.trim()))
    }
    return out.toList
  }

  def do_spark(review_file: String, business_file: String, top_n: Int): Array[List[Any]] = {
    val sc = SparkContext.getOrCreate()
    sc.setLogLevel("ERROR")

    val reviewRDD = sc.textFile(review_file).map(row => format_rdata(row))
    val bizRDD = sc.textFile(business_file).map(row => format_bdata(row))
      .filter(row => row._2 != null)
      .flatMap(row => split_categories(row))

    val zero_value = (0.0, 0.0)
    val seq_fn = (x: Tuple2[Double, Double], y: Float) => ((x._1 + y).toDouble, x._2 + 1.0)
    val comb_fn = (x: Tuple2[Double, Double], y: Tuple2[Double, Double]) => (x._1 + y._1, x._2 + y._2)

    return reviewRDD.join(bizRDD)
      .map(x => (x._2._2, x._2._1))
      .aggregateByKey(zero_value)(seq_fn, comb_fn)
      .mapValues(x => BigDecimal(x._1 / x._2).setScale(1, BigDecimal.RoundingMode.HALF_UP).toDouble)
      .sortBy(x => (-x._2, x._1))
      .take(top_n).map(item => item.productIterator.toList)
  }

  def do_py(review_file: String, business_file: String, top_n: Int): Array[List[Any]] = {
    val review_lines = Source.fromFile(review_file).getLines.map(format_rdata)
    val biz_lines = Source.fromFile(business_file).getLines
      .map(format_bdata)
      .filter(row => row._2 != null)
      .flatMap(row => split_categories(row)).toList

    val rdata = review_lines.toSeq.groupBy(_._1).mapValues(_.map(_._2)).toList
    var rdatadict = scala.collection.mutable.Map[String, Seq[Float]]()

    for (x <- rdata) {
      rdatadict(x._1) = x._2
    }

    var cat_sum = scala.collection.mutable.Map[String, Float]()
    var cat_count = scala.collection.mutable.Map[String, Int]()

    for (row <- biz_lines) {
      val b_id = row._1

      if (rdatadict.contains(b_id)) {
        for (x: Float <- rdatadict(b_id)) {
          if (cat_sum.contains(row._2)) {
            cat_sum(row._2) = cat_sum(row._2) + x
          } else {
            cat_sum(row._2) = x
          }

          if (cat_count.contains(row._2)) {
            cat_count(row._2) = cat_count(row._2) + 1
          } else {
            cat_count(row._2) = 1
          }
        }
      }
    }

    var avg_data = new ArrayBuffer[Tuple2[String, Float]]

    for (k <- cat_sum.keys) {
      avg_data += ((k, cat_sum(k) / cat_count(k).toFloat))
    }

    val output_array = avg_data.map(x => (x._1,
      BigDecimal(x._2.toDouble).setScale(1, BigDecimal.RoundingMode.HALF_UP).toDouble
      ))
      .sortBy(x => (-x._2, x._1)).
      toArray
    return output_array.take(top_n).map(item => item.productIterator.toList)
  }

  def main(args:Array[String]): Unit = {
    for (arg <- args) {
      println(arg)
    }

    val review_file = args(0)
    val business_file = args(1)
    val output_file = args(2)
    val use_spark = if (args(3) == "spark") true else false
    val top_n = args(4).toInt

    val output: Map[String, Any]
      = if (use_spark) Map("result"->do_spark(review_file, business_file, top_n))
        else Map("result"->do_py(review_file, business_file, top_n))

    implicit val formats = org.json4s.DefaultFormats
    val formatted_output = org.json4s.jackson.Serialization.write(output)

    println(formatted_output)

    val out = new PrintWriter(new File(output_file))
    out.write(formatted_output)
    out.close()
  }
}

File: python/task2.py
from pyspark import SparkContext
import json
import sys
import time
from collections import defaultdict

class Task2:
    def __init__(self):
        self.review_file = sys.argv[1]
        self.business_file = sys.argv[2]
        self.output_file = sys.argv[3]
        self.use_spark = True if sys.argv[4] == "spark" else False

        if self.use_spark:
            self.sc = SparkContext.getOrCreate()
            self.sc.setLogLevel("ERROR")

        self.top_n = int(sys.argv[5])

        self.output = {}
    
    def read_data(self):
        if self.use_spark:
            self.review_data = self.sc.textFile(self.review_file)
            self.business_data = self.sc.textFile(self.business_file)
        else:
            with open(self.review_file) as f:
                self.review_data = f.readlines()
            with open(self.business_file) as f:
                self.business_data = f.readlines()

    def get_top_rated_categories(self):
        if self.use_spark:
            self.spark_get_top(self.top_n)
        else:
            self.py_get_top()

    @staticmethod
    def format_review_data(row):
        data = json.loads(row)
        return (data["business_id"], float(data["stars"]))
    
    @staticmethod
    def format_business_data(row):
        data = json.loads(row)
        return (data["business_id"], data["categories"])
    
    @staticmethod
    def split_categories(row):
        return [(row[0], x.strip()) for x in row[1].strip().split(",")]

    def py_get_top(self):
        flat_map = lambda x, fn: [item for lst in x for item in fn(lst)]
        review_data = list(map(Task2.format_review_data, self.review_data))
        business_data = flat_map(filter(lambda row: row[1] is not None, map(Task2.format_business_data, self.business_data)), Task2.split_categories)
        
        # Perform the inner join
        review_dict = defaultdict(list)
        for x in review_data:
            review_dict[x[0]].append(x[1])

        cat_sum = defaultdict(int)
        cat_count = defaultdict(int)

        for row in business_data:
            b_id = row[0]
            if b_id in review_dict:
                for val in review_dict[b_id]:
                    cat_sum[row[1]] += val
                    cat_count[row[1]] += 1

        avg_data = []
        for k in cat_sum:
            avg_data.append((k, cat_sum[k] / cat_count[k]))
        
        avg_data.sort(key=lambda x: (-x[1], x[0]))

        self.output["result"] = list(map(lambda x: (x[0], round(x[1], 1)), avg_data[:self.top_n]))

    
    def spark_get_top(self, top_n):
        r_data = self.review_data.map(lambda row: Task2.format_review_data(row))
        b_data = self.business_data.map(lambda row: Task2.format_business_data(row))\
            .filter(lambda row: row[1] is not None)\
            .flatMap(lambda row: Task2.split_categories(row))

        zero_value = (0, 0)
        seq_fn = (lambda x, y: (x[0] + y, x[1] + 1))
        comb_fn = (lambda x, y: (x[0] + y[0], x[1] + y[1]))
        
        self.output["result"] = r_data.join(b_data)\
                                .map(lambda x: (x[1][1], x[1][0]))\
                                    .aggregateByKey(zero_value, seq_fn, comb_fn)\
                                        .mapValues(lambda x: x[0]/x[1])\
                                            .sortBy(lambda x: (-x[1], x[0]))\
                                                .map(lambda x: (x[0], round(x[1], 1)))\
                                                    .take(top_n)
    
    def write_output(self):
        with open(self.output_file, "w") as f:
            json.dump(self.output, f)

if __name__ == "__main__":
    start_time = time.time()

    runner = Task2()    
    runner.read_data()
    runner.get_top_rated_categories()
    runner.write_output()

    end_time = time.time()

    print("Total time taken: {}".format(end_time - start_time))
